react/src/hooks/__tests__/useFeatureFlags.test.tsx

```diff
@@ -0,0 +1,70 @@
+import React from 'react'
+import { mocked } from 'ts-jest/utils'
+import { renderHook, act } from '@testing-library/react-hooks'
+import posthog from 'posthog-js'
+import { PostHogProvider, FeatureFlags } from '../../context'
+import { useFeatureFlags } from '../'
+
+describe('useFeatureFlags hook', () => {
+    const ACTIVE_FEATURE_FLAGS: FeatureFlags['active'] = ['example_feature_1', 'example_feature_2', 'example_feature_3']
+    const ENABLED_FEATURE_FLAGS: FeatureFlags['enabled'] = {
+        example_feature_1: true,
+        example_feature_2: true,
+        example_feature_3: false,
+    }
+
+    const wrapper = ({ children }: { children: any }) => <PostHogProvider client={posthog}>{children}</PostHogProvider>
+
+    beforeEach(() => {
+        posthog.init('test_token', {
+            api_host: 'https://test.com',
+        })
+    })
+
+    it('should return an empty `enabled` object by default', () => {
+        const mockedPosthog = mocked(posthog)
+        jest.spyOn(mockedPosthog, 'onFeatureFlags').mockImplementationOnce(() => undefined)
+
+        const { result } = renderHook(() => useFeatureFlags(), { wrapper })
+        expect(result.current).toEqual({
+            enabled: {},
+        })
+    })
+
+    it('should return `active` and `enabled` features when feature flags are changed', async () => {
```

Comment:
How to rewrite this test using given2/no mocking:

```js

given('subject', () => renderHook(() => useFeatureFlags(), { wrapper: given.renderProvider }).result)

given('renderProvider', () => ({ children }) => <PostHogProvider client={given.posthog}>{children}</PostHogProvider>)

it('should return `active` and `enabled` features when feature flags are changed', () => {
  given('posthog', () => ({
    onFeatureFlags: (callback) => callback(ACTIVE_FEATURE_FLAGS),
    isFeatureEnabled: (flag) => flag !== 'example_feature_3'
  })
  expect(given.subject.current).toEqual({
       active: ACTIVE_FEATURE_FLAGS,
       enabled: ENABLED_FEATURE_FLAGS,
  })
})
```

Since you can override given at `it` or `describe` level this should result in much less boilerplate overall.

There's ways to cut the repetition down further if given2 is applied to the whole file (e.g. making onFeatureFlags behavior another given).

Reply:
That's great, thank you. This is a much more straight-forward way to mock functions, especially given that they can be overriden. I've followed your advice and set the default mocks at the **describe** level, and override within the **test** only as necessary.